### ISSUE TITLE
Detect and flag as errors forward references to variables

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#6303]
With the old semantic analyzer this code is an error:
```python
def func() -> None:
    x
    x = 1
```
We should also flag it as an error in the new semantic analyzer (ideally also at module scope but it doesn't work currently and there is a separate issue).

This is a follow-up for https://github.com/python/mypy/pull/6280

**Comment by ilevkivskyi** [comment]
This looks like a part of a more general problem with the new analyzer. In the old analyzer definitions are always processed in the order they appear in the source file. In the new analyzer it is no longer true and we need a (robust) way to track order of definitions. So far this caused at least three hard to solve issues:
* Forward references work to good, like in the original example.
* Scopes can be confused for the same names, for example:
  ```python
  X = 1
  class C:
      X = X  # Cannot determine type of 'X'
  ```
* An original definition may not take precedence, see e.g. https://github.com/python/mypy/pull/6696

I therefore raise priority to high.

**Comment by erictraut** [comment]
This appears to no longer be a bug. The latest version of mypy handles these cases correctly.
